experimental/packages/opentelemetry-sdk-metrics-base/src/exemplar/ExemplarReservoir.ts

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ValueType, Attributes } from '@opentelemetry/api-metrics'
+import { Context, HrTime, isSpanContextValid, trace } from '@opentelemetry/api'
+import { Exemplar } from './Exemplar'
+
+
+/**
+ * An interface for an exemplar reservoir of samples.
+ */
+export interface ExemplarReservoir {
+
+  /** Offers a measurement to be sampled. */
+  offer(
+    value: ValueType,
+    timestamp: HrTime,
+    attributes: Attributes,
+    ctx: Context    
+  ): void;
+  /**
+   * Returns accumulated Exemplars and also resets the reservoir
+   * for the next sampling period
+   * 
+   * @param pointAttributes The attributes associated with metric point. 
+   */
+  collect(pointAttributes: Attributes): Exemplar[];
+}
+
+
+class StorageItem {
```

Comment:
The name `StorageItem` sounds vague to me. Can we name it with a more meaningful name like `ReservoirCell` or `ExamplarBucket` (the name Bucket is used in `FixedSizeExemplarReservoirBase` already) etc. to indicate it is a pre-allocated storage item for exemplars in ExemplarReservoir?